fondant/dataset.py

```diff
@@ -76,60 +99,129 @@ def load_dataframe(self, spec: FondantComponentSpec) -> dd.DataFrame:
 
         return df
 
-    def _upload_index(self, df: dd.DataFrame):
-        # get remote path
-        remote_path = self.manifest.index.location
-
-        # upload to the cloud
-        dd.to_parquet(
-            df,
-            remote_path,
-            schema=self.index_schema,
-            overwrite=True,
+    @staticmethod
+    def _create_write_dataframe_task(
+        *, df: dd.DataFrame, remote_path: str, schema: t.Dict[str, str]
+    ) -> dd.core.Scalar:
+        """
+        Creates a delayed Dask task to upload the given DataFrame to the remote storage location
+         specified in the manifest.
+
+        Args:
+            df: The DataFrame to be uploaded.
+            remote_path: the location to upload the subset to
+            schema: the schema of the dataframe to write
+
+        Returns:
+             A delayed Dask task that uploads the DataFrame to the remote storage location when
+              executed.
+        """
+
+        # Define task to upload index to remote storage
+        write_task = dd.to_parquet(
+            df, remote_path, schema=schema, overwrite=False, compute=False
         )
-
-    def _upload_subset(
-        self, name: str, fields: t.Mapping[str, Field], df: dd.DataFrame
-    ):
-        # create expected schema
-        expected_schema = {field.name: field.type.name for field in fields.values()}
-        expected_schema.update(self.index_schema)
-
-        # get remote path
-        remote_path = self.manifest.subsets[name].location
-        # upload to the cloud
-        dd.to_parquet(df, remote_path, schema=expected_schema, overwrite=True)
+        return write_task
 
     def write_index(self, df: dd.DataFrame):
+        """
+        Write the index dataframe to a remote location.
+        Args:
+            df: The output Dask dataframe returned by the user.
+        """
+        remote_path = self.manifest.index.location
         index_columns = list(self.manifest.index.fields.keys())
 
         # load index dataframe
         index_df = df[index_columns]
```

Comment:
Does this `index_df` have an index ? Maybe we need to call `.set_index()` ?

Reply:
we're using both the `id` and `source` as index. It does not seem like dask supports multiindex though https://dask.discourse.group/t/everything-about-multiindex-in-dask/593/2

Reply:
We could just index `id` since it is better then no index (this is also what happens in the test data see split.py). I'll create a ticket to solve multi-index.